File: sdk/typescript/tests/shell.test.mjs
import assert from "node:assert/strict";
import { appendFile, copyFile, mkdtemp, rm, stat } from "node:fs/promises";
import { tmpdir } from "node:os";
import { basename, join } from "node:path";
import test from "node:test";

import {
  ShellExecutionError,
  ShellExecutor,
  ShellNotAllowedError,
  ShellProtocolError,
  ShellReplayError,
} from "../dist/shell.js";

class FakeClient {
  constructor(verdict = "ALLOW", error, after) {
    this.verdict = verdict;
    this.error = error;
    this.after = after;
    this.actions = [];
  }

  async decide(action) {
    this.actions.push(action);
    await this.after?.();
    if (this.error !== undefined) throw this.error;
    return { decision: this.verdict, allowed: this.verdict === "ALLOW" };
  }
}

test("structured command is allowed and normalized without leaking input", async () => {
  const client = new FakeClient();
  const executor = new ShellExecutor(client);
  const code = `let d="";process.stdin.setEncoding("utf8");process.stdin.on("data",c=>d+=c);process.stdin.on("end",()=>process.stdout.write(process.argv.slice(1).join(",")+"|"+process.env.LATCH_VALUE+"|"+d));`;
  const result = await executor.run({
    executionId: "exec_node_001",
    executable: process.execPath,
    args: ["-e", code, "one", "two"],
    env: { LATCH_VALUE: "secret-value" },
    stdin: "payload",
  });
  assert.equal(result.exitCode, 0);
  assert.equal(Buffer.from(result.stdout).toString("utf8"), "one,two|secret-value|payload");
  const action = client.actions[0];
  assert.equal(action.tool, "shell.exec");
  assert.equal(action.metadata.surface, "argv");
  const encoded = JSON.stringify(action);
  assert.equal(encoded.includes("secret-value"), false);
  assert.equal(encoded.includes("payload"), false);
  assert.ok(action.arguments.environment_sha256);
  assert.ok(action.arguments.stdin_sha256);
});

test("BLOCK, REQUIRE_APPROVAL, and outage start no process", async (t) => {
  for (const [name, client, ErrorType] of [
    ["block", new FakeClient("BLOCK"), ShellNotAllowedError],
    ["approval", new FakeClient("REQUIRE_APPROVAL"), ShellNotAllowedError],
    ["outage", new FakeClient("ALLOW", new Error("offline")), Error],
  ]) {
    await t.test(name, async () => {
      const directory = await mkdtemp(join(tmpdir(), "latch-shell-"));
      const marker = join(directory, "started");
      try {
        const code = `require("node:fs").writeFileSync(process.argv[1],"started")`;
        await assert.rejects(
          new ShellExecutor(client).run({ executionId: "exec_denied_001", executable: process.execPath, args: ["-e", code, marker] }),
          ErrorType,
        );
        await assert.rejects(stat(marker), { code: "ENOENT" });
      } finally {
        await rm(directory, { recursive: true, force: true });
      }
    });
  }
});

test("failed process remains consumed and replay is rejected", async () => {
  const client = new FakeClient();
  const executor = new ShellExecutor(client);
  const command = { executionId: "exec_replay_001", executable: process.execPath, args: ["-e", "process.exit(7)"] };
  await assert.rejects(executor.run(command), ShellExecutionError);
  await assert.rejects(executor.run(command), ShellReplayError);
  assert.equal(client.actions.length, 1);
});

test("timeout and bounded output stop allowed processes", async () => {
  const timeoutExecutor = new ShellExecutor(new FakeClient(), { timeoutMs: 50 });
  await assert.rejects(
    timeoutExecutor.run({ executionId: "exec_timeout_001", executable: process.execPath, args: ["-e", "setTimeout(()=>{},2000)"] }),
    ShellExecutionError,
  );
  const outputExecutor = new ShellExecutor(new FakeClient(), { maxOutputBytes: 1024 });
  await assert.rejects(
    outputExecutor.run({ executionId: "exec_output_001", executable: process.execPath, args: ["-e", "process.stdout.write('x'.repeat(4096))"] }),
    (error) => error instanceof ShellExecutionError && error.result.stdout.byteLength === 1024,
  );
});

test("changed executable and hostile environment objects fail before start", async () => {
  const directory = await mkdtemp(join(tmpdir(), "latch-shell-"));
  const copied = join(directory, basename(process.execPath));
  try {
    await copyFile(process.execPath, copied);
    const client = new FakeClient("ALLOW", undefined, () => appendFile(copied, "changed"));
    await assert.rejects(
      new ShellExecutor(client).run({ executionId: "exec_changed_001", executable: copied }),
      ShellProtocolError,
    );
  } finally {
    await rm(directory, { recursive: true, force: true });
  }

  const conflictClient = new FakeClient();
  await assert.rejects(
    new ShellExecutor(conflictClient).run({
      executionId: "exec_env_conflict",
      executable: process.execPath,
      env: { LATCH_VALUE: "one" },
      unsetEnv: ["LATCH_VALUE"],
    }),
    ShellProtocolError,
  );
  assert.equal(conflictClient.actions.length, 0);

  let getterRead = false;
  const hostile = {};
  Object.defineProperty(hostile, "LATCH_VALUE", { enumerable: true, get() { getterRead = true; return "secret"; } });
  await assert.rejects(
    new ShellExecutor(new FakeClient()).run({ executionId: "exec_getter_001", executable: process.execPath, env: hostile }),
    ShellProtocolError,
  );
  assert.equal(getterRead, false);

  let argumentRead = false;
  const hostileArgs = [];
  Object.defineProperty(hostileArgs, 0, { enumerable: true, get() { argumentRead = true; return "--version"; } });
  hostileArgs.length = 1;
  await assert.rejects(
    new ShellExecutor(new FakeClient()).run({ executionId: "exec_arg_getter", executable: process.execPath, args: hostileArgs }),
    ShellProtocolError,
  );
  assert.equal(argumentRead, false);
});

test("explicit shell text uses the separate protected surface", async () => {
  const client = new FakeClient();
  const result = await new ShellExecutor(client).runShell({ executionId: "exec_script_001", command: "echo shell-ok" });
  assert.match(Buffer.from(result.stdout).toString("utf8"), /shell-ok/);
  assert.equal(client.actions[0].arguments.command, "echo shell-ok");
  assert.equal(client.actions[0].metadata.surface, "shell");
});

File: sdk/go/latch/shell_test.go
package latch

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	api "github.com/princebabou/Latch/pkg/api/v1"
	"github.com/princebabou/Latch/pkg/models"
)

type fakeShellClient struct {
	actions  []api.Action
	decision models.Decision
	err      error
	after    func()
}

func (client *fakeShellClient) Decide(_ context.Context, action api.Action) (api.DecisionResponse, error) {
	client.actions = append(client.actions, action)
	if client.after != nil {
		client.after()
	}
	if client.err != nil {
		return api.DecisionResponse{}, client.err
	}
	decision := client.decision
	if decision == "" {
		decision = models.DecisionAllow
	}
	return api.DecisionResponse{Decision: decision}, nil
}

func TestShellHelperProcess(t *testing.T) {
	if os.Getenv("LATCH_SHELL_HELPER") != "1" {
		return
	}
	if marker := os.Getenv("LATCH_SHELL_MARKER"); marker != "" {
		_ = os.WriteFile(marker, []byte("started"), 0o600)
	}
	mode := os.Getenv("LATCH_SHELL_MODE")
	switch mode {
	case "sleep":
		time.Sleep(2 * time.Second)
	case "output":
		fmt.Print(strings.Repeat("x", 4096))
	default:
		input, _ := io.ReadAll(os.Stdin)
		separator := len(os.Args)
		for index, argument := range os.Args {
			if argument == "--" {
				separator = index + 1
				break
			}
		}
		fmt.Printf("%s|%s|%s", strings.Join(os.Args[separator:], ","), os.Getenv("LATCH_SHELL_VALUE"), input)
	}
	os.Exit(0)
}

func TestShellStructuredCommandAllowedAndNormalized(t *testing.T) {
	client := &fakeShellClient{}
	executor, err := NewShellExecutor(client)
	if err != nil {
		t.Fatal(err)
	}
	result, err := executor.Run(context.Background(), ShellCommand{
		ExecutionID: "exec_allowed_001", Executable: os.Args[0],
		Args: []string{"-test.run=TestShellHelperProcess", "--", "one", "two"},
		Environment: map[string]string{
			"LATCH_SHELL_HELPER": "1", "LATCH_SHELL_VALUE": "secret-value",
		},
		Stdin: "payload",
	})
	if err != nil {
		t.Fatal(err)
	}
	if result.ExitCode != 0 || string(result.Stdout) != "one,two|secret-value|payload" {
		t.Fatalf("result = %#v", result)
	}
	if len(client.actions) != 1 {
		t.Fatalf("actions = %d", len(client.actions))
	}
	action := client.actions[0]
	if action.Tool != "shell.exec" || action.Operation != "execute" || action.Metadata["surface"] != "argv" {
		t.Fatalf("action = %#v", action)
	}
	if action.Resource == "" || !filepath.IsAbs(action.Resource) {
		t.Fatalf("resource = %q", action.Resource)
	}
	encoded, err := json.Marshal(action)
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(string(encoded), "secret-value") || strings.Contains(string(encoded), "payload") {
		t.Fatalf("sensitive process data leaked into action: %s", encoded)
	}
	if action.Arguments["environment_sha256"] == "" || action.Arguments["stdin_sha256"] == "" {
		t.Fatalf("missing snapshot digests: %#v", action.Arguments)
	}
}

func TestShellDenialAndUnavailableStartNothing(t *testing.T) {
	for _, test := range []struct {
		name           string
		client         *fakeShellClient
		wantNotAllowed bool
	}{
		{"block", &fakeShellClient{decision: models.DecisionBlock}, true},
		{"approval", &fakeShellClient{decision: models.DecisionRequireApproval}, true},
		{"unavailable", &fakeShellClient{err: errors.New("offline")}, false},
	} {
		t.Run(test.name, func(t *testing.T) {
			marker := filepath.Join(t.TempDir(), "started")
			executor, _ := NewShellExecutor(test.client)
			_, err := executor.Run(context.Background(), ShellCommand{
				ExecutionID: "exec_denied_001", Executable: os.Args[0],
				Args:        []string{"-test.run=TestShellHelperProcess"},
				Environment: map[string]string{"LATCH_SHELL_HELPER": "1", "LATCH_SHELL_MARKER": marker},
			})
			if err == nil {
				t.Fatal("expected error")
			}
			var notAllowed *ShellNotAllowedError
			if test.wantNotAllowed && !errors.As(err, &notAllowed) {
				t.Fatalf("error = %T %v, want ShellNotAllowedError", err, err)
			}
			if _, statErr := os.Stat(marker); !os.IsNotExist(statErr) {
				t.Fatalf("process started: %v", statErr)
			}
		})
	}
}

func TestShellReplayAndFailureStayConsumed(t *testing.T) {
	client := &fakeShellClient{}
	executor, _ := NewShellExecutor(client, WithShellTimeout(50*time.Millisecond))
	command := ShellCommand{
		ExecutionID: "exec_replay_001", Executable: os.Args[0],
		Args:        []string{"-test.run=TestShellHelperProcess"},
		Environment: map[string]string{"LATCH_SHELL_HELPER": "1", "LATCH_SHELL_MODE": "sleep"},
	}
	_, firstErr := executor.Run(context.Background(), command)
	var executionError *ShellExecutionError
	if !errors.As(firstErr, &executionError) {
		t.Fatalf("first error = %T %v", firstErr, firstErr)
	}
	_, secondErr := executor.Run(context.Background(), command)
	var replayError *ShellReplayError
	if !errors.As(secondErr, &replayError) || len(client.actions) != 1 {
		t.Fatalf("second error/actions = %v / %d", secondErr, len(client.actions))
	}
}

func TestShellRejectsExecutableChangeAfterApproval(t *testing.T) {
	copyPath := filepath.Join(t.TempDir(), "helper")
	if runtime.GOOS == "windows" {
		copyPath += ".exe"
	}
	payload, err := os.ReadFile(os.Args[0])
	if err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(copyPath, payload, 0o700); err != nil {
		t.Fatal(err)
	}
	marker := filepath.Join(t.TempDir(), "started")
	client := &fakeShellClient{after: func() {
		file, openErr := os.OpenFile(copyPath, os.O_APPEND|os.O_WRONLY, 0)
		if openErr == nil {
			_, _ = file.Write([]byte("changed"))
			_ = file.Close()
		}
	}}
	executor, _ := NewShellExecutor(client)
	_, err = executor.Run(context.Background(), ShellCommand{
		ExecutionID: "exec_changed_01", Executable: copyPath,
		Args:        []string{"-test.run=TestShellHelperProcess"},
		Environment: map[string]string{"LATCH_SHELL_HELPER": "1", "LATCH_SHELL_MARKER": marker},
	})
	var protocolError *ShellProtocolError
	if !errors.As(err, &protocolError) {
		t.Fatalf("error = %T %v", err, err)
	}
	if _, statErr := os.Stat(marker); !os.IsNotExist(statErr) {
		t.Fatalf("changed executable started: %v", statErr)
	}
}

func TestShellOutputLimitAndEnvironmentConflicts(t *testing.T) {
	client := &fakeShellClient{}
	executor, _ := NewShellExecutor(client, WithShellMaxOutputBytes(1024))
	_, err := executor.Run(context.Background(), ShellCommand{
		ExecutionID: "exec_output_001", Executable: os.Args[0],
		Args:        []string{"-test.run=TestShellHelperProcess"},
		Environment: map[string]string{"LATCH_SHELL_HELPER": "1", "LATCH_SHELL_MODE": "output"},
	})
	var executionError *ShellExecutionError
	if !errors.As(err, &executionError) || len(executionError.Result.Stdout) != 1024 {
		t.Fatalf("error/result = %v / %#v", err, executionError)
	}

	conflictClient := &fakeShellClient{}
	conflictExecutor, _ := NewShellExecutor(conflictClient)
	_, err = conflictExecutor.Run(context.Background(), ShellCommand{
		ExecutionID: "exec_env_conflict", Executable: os.Args[0],
		Environment: map[string]string{"LATCH_VALUE": "one"}, UnsetEnvironment: []string{"LATCH_VALUE"},
	})
	var protocolError *ShellProtocolError
	if !errors.As(err, &protocolError) || len(conflictClient.actions) != 0 {
		t.Fatalf("error/actions = %v / %d", err, len(conflictClient.actions))
	}
}

func TestShellExplicitScript(t *testing.T) {
	client := &fakeShellClient{}
	executor, _ := NewShellExecutor(client)
	result, err := executor.RunShell(context.Background(), ShellScript{
		ExecutionID: "exec_script_001", Command: "echo shell-ok",
	})
	if err != nil {
		t.Fatal(err)
	}
	if result.ExitCode != 0 || !strings.Contains(string(result.Stdout), "shell-ok") {
		t.Fatalf("result = %#v", result)
	}
	if client.actions[0].Arguments["command"] != "echo shell-ok" || client.actions[0].Metadata["surface"] != "shell" {
		t.Fatalf("action = %#v", client.actions[0])
	}
}
